test/systemtest/earlgrey/test_sim_verilator.py

```diff
@@ -0,0 +1,164 @@
+        assert self.uart0_device_path.is_char_device()
+        self._log.debug("Found uart0 device file at {}".format(
+            str(self.uart0_device_path)))
+
+        self.uart0_log_path = self._work_dir / 'uart0.log'
+        assert self.uart0_log_path.is_file()
+
+        # SPI
+        spi0_match = self.p_sim.find_in_output(
+            re.compile(r'SPI: Created (/dev/pts/\d+) for spi0\.'), 1)
+        assert spi0_match is not None
+        self.spi0_device_path = Path(spi0_match.group(1))
+        assert self.spi0_device_path.is_char_device()
+        self._log.debug("Found spi0 device file at {}".format(
+            str(self.spi0_device_path)))
+
+        self.spi0_log_path = self._work_dir / 'spi0.log'
+        assert self.spi0_log_path.is_file()
+
+        # GPIO
+        self.gpio0_fifo_write_path = self._work_dir / 'gpio0-write'
+        assert self.gpio0_fifo_write_path.is_fifo()
+
+        self.gpio0_fifo_read_path = self._work_dir / 'gpio0-read'
+        assert self.gpio0_fifo_read_path.is_fifo()
+
+        self._log.info("Simulation startup completed.")
+
+    def uart0(self):
+        if self._uart0 is None:
+            log_dir_path = self._work_dir / 'uart0'
+            log_dir_path.mkdir()
+            log.info("Opening UART on device {} ({} baud)".format(
+                str(self.uart0_device_path), self.UART0_SPEED))
+            self._uart0 = utils.LoggingSerial(
+                str(self.uart0_device_path),
+                self.UART0_SPEED,
+                timeout=1,
+                log_dir_path=log_dir_path,
+                default_filter_func=utils.filter_remove_device_sw_log_prefix)
+
+        return self._uart0
+
+    def terminate(self):
+        """ Gracefully terminate the simulation """
+        self._uart0.close()
+
+        self.p_sim.send_ctrl_c()
+
+        # Give the process some time to clean up
+        self.p_sim.proc.wait(timeout=5)
+        assert self.p_sim.proc.returncode == 0
+
+        try:
+            self.p_sim.terminate()
+        except ProcessLookupError:
+            # process is already dead
+            pass
+
+
+@pytest.fixture(params=config.TEST_APPS_SELFCHECKING)
+def app_selfchecking_elf(request, bin_dir):
+    """ A self-checking device application as ELF for Verilator simulation """
+
+    test_filename = request.param + '_sim_verilator.elf'
+    bin_path = bin_dir / 'sw/device/tests' / test_filename
+    assert bin_path.is_file()
+    return bin_path
+
+
+def test_apps_selfchecking(tmp_path, bin_dir, app_selfchecking_elf):
+    """
+    Run a self-checking application on a Earl Grey Verilator simulation
+
+    The ROM is initialized with the default boot ROM, the flash is initialized
+    with |app_selfchecking_elf|.
+
+    Self-checking applications are expected to return PASS or FAIL in the end.
+    """
+
+    sim_path = bin_dir / "hw/top_earlgrey/Vtop_earlgrey_verilator"
+    rom_elf_path = bin_dir / "sw/device/boot_rom/boot_rom_sim_verilator.elf"
```

Comment:
Do we want to hard-code these? Especially if we end up with multiple tops to support different FPGA builds we may want to be able to run self checking tests across verilator simulations of the different tops so we'd have a changing `sim_path for instance.. We could write separate `test_X` functions to deal with this but seems poor to copy-paste the same code over and over.

Perhaps we could refactor this into a separate function that takes sim_path and rom_elf_path as parameters which is then called by `test_apps_selfchecking`? Then it's easy to extend if we do want to run tests across multiple different verilator binaries whilst maintaining the 'opinionated' style.

Reply:
> Do we want to hard-code these?

I think yes. This test is specific to the earl grey verilator simulation, as we have it today. The `earlgrey/test_sim_verilator.py` file is therefore, in my mind, the right place for all knowledge about this specific simulation.

In the future, we know that the structuring of our toplevels is likely to change. But we don't know how it will change. When that happens and we add code for the next toplevel or simulation, we should/must think about how to avoid duplicated code. I can imagine, for example, factoring out the `VerilatorSimEarlgrey` class and having a common base class for verilator simulations with our common framework -- or not, if we end up having significantly different peripherals in there.

One question remains though: *where* in `earlgrey/test_sim_verilator.py` should we hard-code values? Currently, knowledge is a bit split between the `VerilatorSimEarlgrey` class (knowledge about how the verilator simulation behaves) and `test_apps_selfchecking()` (file names). I'm not super happy with that, but this split makes `VerilatorSimEarlgrey` reusable and avoids making it a fixture itself (which comes with its own chain of downsides). Let me know if you can think of a better way of structuring this code.

Reply:
Please don't hard-code the ROM elf path, we're working on two ROMs in parallel (`sw/device/boot_rom` is what everyone uses now, but we'll soon be using `sw/device/mask_rom` as we start testing the production Mask ROM).

Reply:
That's fine (and I anticipated that). Once the mask rom appears we'll adjust the test to make us of it (or both ROMs). The tests probe for specifics of the ROM output, so simply having a config file with the ELF paths doesn't do the trick -- the test will need to be adjusted.

Reply:
Ah, ok, so we'd end up with a different test subclass, I guess, where we can override the fixture there. That sounds good, and might make it more obvious which tests go with which ROM.